custom/icds_reports/management/commands/split_j_and_k_state.py

```diff
@@ -0,0 +1,110 @@
+import copy
+import json
+import subprocess
+
+import os
+from django.core.management import BaseCommand
+
+from custom.icds_reports.utils.topojson_util.topojson_util import (
+    get_topojson_directory,
+    copy_custom_metadata,
+    get_state_v3_topojson_file,
+    get_district_v3_topojson_file
+)
+
+
+J_AND_K = 'J&K'
+LADAKH = 'Ladakh'
+
+
+class Command(BaseCommand):
+    help = (
+        "Split Jammu and Kashmir and Ladakh into separate states, as per "
+        "https://en.wikipedia.org/wiki/Jammu_and_Kashmir_Reorganisation_Act,_2019."
+    )
+    # this command was used for https://app.asana.com/0/1112385193248823/1157605674172491
+
+    def handle(self, *args, **kwargs):
+        self.input_dir = get_topojson_directory()
+        self._update_state_file()
+        self._update_district_file()
+
+    def _update_state_file(self):
+        # loading state topojson object
+        state_topojson_file = get_state_v3_topojson_file()
+        state_topojson = state_topojson_file.topojson
+
+        # remove J&K from list of geometries
+        geometries = state_topojson['objects']['ind']['geometries']
+        new_geometries = [g for g in geometries if g['id'] != 'J&K']
+        state_topojson['objects']['ind']['geometries'] = new_geometries
+
+        # save a new file
+        tmp_state_filename = os.path.join(self.input_dir, 'states_v4_tmp.topojson')
+        with open(tmp_state_filename, 'w+') as new_map_file:
+            new_map_file.write(json.dumps(state_topojson))
+
+        # assumes these files are in the input directory.
+        # get them from  https://app.asana.com/0/1112385193248823/1157605674172491
+        j_k_file = os.path.join(self.input_dir, 'Jammu_and_Kashmir_State.shp')
+        ladakh_file = os.path.join(self.input_dir, 'Ladakh_State.shp')
+
+        new_state_filename = os.path.join(self.input_dir, 'states_v4.topojson')
+
+        # now we merge in the new shape files using mapshaper : https://www.npmjs.com/package/mapshaper
+        # see https://gis.stackexchange.com/a/221075/126250 for details
+        mapshaper_command = f"""mapshaper \
+          -i {tmp_state_filename} {j_k_file} {ladakh_file} snap combine-files \
+          -rename-layers states,jk,ladakh \
+          -merge-layers target=states,jk,ladakh name=ind force \
+          -o {new_state_filename}
+        """
+        subprocess.call(mapshaper_command, shell=True)
+
+        # now open the newly created file
+        with open(new_state_filename, 'r') as f:
+            new_states = json.loads(f.read())
+
+        # ...and add metadata back
+        copy_custom_metadata(state_topojson, new_states)
+
+        # we also have to manually populate metadata for the two states
+        jk = new_states['objects']['ind']['geometries'][-2]
+        jk['id'] = J_AND_K
+        jk['properties'] = {"State": "37", "name": J_AND_K}
```

Comment:
I made up these numbers based on what the previous highest State ID in the file was. Pretty sure they aren't used for anything.

Reply:
If these are not used, this seems fine, otherwise I would have expected those to match the codes used in locations and the name of the state.

Reply:
Interesting. Actually looking closer it does seem like the `site_code` field on prod matches what's in the topojson file so I'll stick with that for J&K. Ladakh has not been created in the location hierarchy yet, so I think that one can stay.

f79012b521baf3d8c600d46fc61e1fa693759322